devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
+
+- `get` requests
+- `pop` requests
+
+The consumer provides a `JOB_REQUEST_FUNCTION` that is executed when that request is executed. If the `JOB_REQUEST_FUNCTION` returns `true`, that job is considered completed, and it's  `JOB_COMPLETE_CALLBACK` is called. Once a job is completed, any new sessions that will be started will not see the completed job.
+
+Even after a job is completed, it is not discarded. A job is discarded when all the sessions that were active when the job was pushed have popped it. If all the sessions that were active when the job was queued have popped a job without indicating that the job is complete, the `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_ERROR` as the result.
+
+Sessions make progress independent of each other. It is possible that one session has popped all the queued requests, whereas another session has not popped any. A session can be thought of as an independent "view" of the job queue. Once a session has called `pop` on a job, it cannot call `pop` or `get` on the same job again. If a session has called `get` on a job, it can call `pop` on that job but cannot call `get` on that `job`.
+
+A producer of jobs must call `push` to queue jobs. All jobs are queued in a single queue and executed in FIFO order. The producer provides a `JOB_COMPLETE_CALLBACK` to be called when the job is completed. If a job is completed because a request marked it as completed (by returning `true` from `JOB_REQUEST_FUNCTION`), `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_OK` as result. In any other case (session ends before job can be completed, job_queue is closing etc), the `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_ERROR`.
```

Comment:
>If a job is completed because a request marked it as completed (by returning `true` from `JOB_REQUEST_FUNCTION`), `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_OK` as result [](http://example.com/codeflow?start=205&length=188)

I'm confused again here, earlier it sounded like each consumer would see the same job, but this says that as soon as one consumer completes the job, then the job complete callback is called. I assume the job_complete_callback is called exactly once.